packages/labs/router/src/routes.ts

```diff
@@ -0,0 +1,212 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {ReactiveController, ReactiveControllerHost} from 'lit';
+
+export interface StringRouteConfig {
+  name?: string | undefined;
+  path: string;
+  render: (params: {[key: string]: string}) => unknown;
+}
+
+export interface URLPatternRouteConfig {
+  name?: string | undefined;
+  pattern: URLPattern;
+  render: (params: {[key: string]: string}) => unknown;
+}
+
+export type RouteConfig = StringRouteConfig | URLPatternRouteConfig;
+
+/**
+ * A reactive controller that performs location-based routing using a
+ * configuration of URL patterns and associated render callbacks.
+ */
+export class Routes implements ReactiveController {
+  protected _host: ReactiveControllerHost & HTMLElement;
```

Comment:
Could we make some of these `readonly`? (And do we care about doing that as a rule?)